scripts/install

```diff
@@ -13,82 +36,147 @@ set -eu
+#=================================================
 
-CHECK_USER "$admin"
+final_path=/var/www/$app
+test ! -e "$final_path" || ynh_die "This path already contains a folder"
 
-CHECK_PATH
+# Normalize the url path syntax
+path_url=$(ynh_normalize_url_path $path_url)
 
-CHECK_DOMAINPATH
+# Check web path availability
+ynh_webpath_available $domain $path_url
+# Register (book) web path
+ynh_webpath_register $app $domain $path_url
 
-CHECK_FINALPATH
+#=================================================
+# STORE SETTINGS FROM MANIFEST
+#=================================================
 
-# Save app settings
 ynh_app_setting_set $app domain $domain
-ynh_app_setting_set $app path $path
+ynh_app_setting_set $app path $path_url
 ynh_app_setting_set $app admin $admin
 ynh_app_setting_set $app is_public $is_public
+ynh_app_setting_set $app language $language
+
+#=================================================
+# STANDARD MODIFICATIONS
+#=================================================
+# DOWNLOAD, CHECK AND UNPACK SOURCE
+#=================================================
+
+ynh_app_setting_set $app final_path $final_path
+# Download, check integrity, uncompress and patch the source from app.src
+ynh_setup_source "$final_path"
+
+#=================================================
+# NGINX CONFIGURATION
+#=================================================
 
-# Create system user dedicace for this app
+# Create a dedicated nginx config
+ynh_add_nginx_config
+
+#=================================================
+# CREATE DEDICATED USER
+#=================================================
+
+# Create a system user
 ynh_system_user_create $app
 
-# Modify dokuwiki conf
-sed -i "s@YNH_ADMIN_USER@$admin@g" ../conf/dokuwiki.php
+#=================================================
+# PHP-FPM CONFIGURATION
+#=================================================
 
-# Copy files to the right place
-sudo mkdir "$final_path"
-ynh_app_setting_set $app final_path $final_path
+# Create a dedicated php-fpm config
+ynh_add_fpm_config
+
+#=================================================
+# SPECIFIC SETUP
+#=================================================
+# CUSTOMIZE DOKUWIKI
+#=================================================
+
+# Set the "language"
+ynh_replace_string "__YNH_LANGUAGE__" "$language" "../conf/local.php"
+
+# Set the "admin" user
+ynh_replace_string "__YNH_ADMIN_USER__" "$admin" "../conf/local.php"
+
+# Copy Yunohost specific configuration
+cp ../conf/local.php $final_path/conf
+
+
+# Restrict user rights by enforcing "read-only" mode for all users
+# See https://www.dokuwiki.org/acl#background_info
+# Default is "8"
+cp ../conf/acl.auth.php $final_path/conf
+
+#=================================================
+# STORE THE CHECKSUM OF THE CONFIG FILE
+#=================================================
+
+# Calculate and store the config file checksum into the app settings
+#ynh_store_file_checksum "$final_path/conf/local.php"
```

Comment:
Why not ?

Reply:
In fact, I do not understand what this helper does. I looked in these places :
- Yunohost documentation : https://yunohost.org/#/contributordoc_fr   
- Example package : https://github.com/YunoHost/example_ynh
- https://helpers.yunohost.org/

Some page does not work on the documentation. What can I do about it ?
https://yunohost.org/#/packaging_apps_scripts_fr

Reply:
This helper will store a checksum of a file, especially a config file, so we can know during the upgrade script if the file have been modified since the installation.
In the upgrade script, the helper `ynh_backup_if_checksum_is_different` will be used to do a backup of the file if the user has modified it.

Reply:
My wording was badly chosen :/ Same as above, I understand what it does as the name is pretty self explanatory. But I don't get why I should need it for the package.

Why a user will ever modify this config file? Well, while writing this, I get it ... If a user go to the admin panel inside Dokuwiki to change some settings like language or plugins, the file is overwritten while saving. So if while upgrading as I planned, the file is copied from conf folder, changes are lost.

So, new question : How can I know that the user has not deleted settings about the LDAP authentication of Yunohost **while** running the upgrade script then ? Have you ever seen this user case?

Reply:
The LDAP config is part of the same config file.
The policy is usually to overwrite the config file by the one from the upgrade script. That's why we're using `ynh_backup_if_checksum_is_different` to do a backup of the previous config if there was a modification.
Are you sure that dokuwiki will modify also the file `local.php` ?

Reply:
>  Are you sure that dokuwiki will modify also the file local.php ?
Yes. The file is modified when an admin goes in the admin panel and changes a settings like the language (switching en => fr)

Tested on a VM
**Before changing language**

```
<?php
/**
 * This is an example of how a local.php could look like.
 * Simply copy the options you want to change from dokuwiki.php
 * to this file and change them.
 *
 * When using the installer, a correct local.php file be generated for
 * you automatically.
 */


/* Basic Settings */
$conf['lang']        = 'en';              //your language

/* Authentication Settings */
$conf['useacl']      = 1;                //Use Access Control Lists to restrict access?
$conf['authtype']    = 'authldap';       //which authentication backend should be used
$conf['passcrypt']   = 'sha1';           //Used crypt method (smd5,md5,sha1,ssha,crypt,mysql,my411)
$conf['superuser']   = 'Gofannon'; //The admin can be user or @group or comma separated list user1,@group1,user2
$conf['manager']     = 'Gofannon'; //The manager can be user or @group or comma separated list user1,@group1,user2

/* LDAP Yunohost config */
$conf['plugin']['authldap']['server']      = 'localhost';
$conf['plugin']['authldap']['port']        = 389;
$conf['plugin']['authldap']['version']     = 3;
$conf['plugin']['authldap']['usertree']    = 'ou=users,dc=yunohost,dc=org';
$conf['plugin']['authldap']['userfilter']  = '(&(uid=%{user})(objectClass=posixAccount))';
# no groups
#$conf['plugin']['authldap']['grouptree']   = 'ou=Group, dc=server, dc=tld';
#$conf['plugin']['authldap']['groupfilter'] = '(&(objectClass=posixGroup)(|(gidNumber=%{gid})(memberUID=%{user})))';

/* Advanced Settings */
$conf['updatecheck'] = 0;                //automatically check for new releases?
$conf['cssdatauri']  = 512;              //Maximum byte size of small images to embed into CSS, won't work on IE<8
```

**After changing language**
```
<?php
/*
 * Dokuwiki's Main Configuration File - Local Settings
 * Auto-generated by config plugin
 * Run for user: himura
 * Date: Wed, 27 Jun 2018 00:08:01 +0200
 */

$conf['lang'] = 'fr';
$conf['useacl'] = 1;
$conf['authtype'] = 'authldap';
$conf['passcrypt'] = 'sha1';
$conf['superuser'] = 'Gofannon';
$conf['manager'] = 'Gofannon';
$conf['updatecheck'] = 0;
$conf['plugin']['authldap']['server'] = 'localhost';
$conf['plugin']['authldap']['usertree'] = 'ou=users,dc=yunohost,dc=org';
$conf['plugin']['authldap']['userfilter'] = '(&(uid=%{user})(objectClass=posixAccount))';
$conf['plugin']['authldap']['version'] = 3;

// end auto-generated content
```

From https://www.dokuwiki.org/config
> The available settings are listed below. Their default values are stored in a file called conf/dokuwiki.php. If you want to modify a setting you should do this in a file called conf/local.php – this makes sure your special configs are not overwritten by a later upgrade. > The config manager will use this file as well. The file conf/local.protected.php can be used for configuration settings which cannot be modified.

It might be interesting to put Yunohost LDAP config in `conf/local.protected.php` then but it will do a second file to think about. Do it worth the hassle ?

Reply:
If it could be done easily, yes it worth it.
The ldap config is an important part of the app, it could be better if the upgrade script can rewrite it during the upgrade. Like that, we now that the upgrade script can fix any ldap issues.

But, there's another problem for me. If we're sure that the local.php file will be rewrite by the admin interface of dokuwiki, in that case I think that we shouldn't overwrite this file at each upgrade. That would means overwrite any user config.

Reply:
> If it could be done easily, yes it worth it.
The ldap config is an important part of the app, it could be better if the upgrade script can rewrite it during the upgrade. Like that, we now that the upgrade script can fix any ldap issues.

Could be done but need testing again.
Settings should be split in two files
1. Settings that can be modified from Dokuwiki : `local.php`
    1. "language" or "admin user" ? If my understand is correct, these settings are stored inside Yunohost so they should not me modified manually by the user. Correct?
2. Settings that cannot be modified : `local.protected.php`
    1. Everything else

> But, there's another problem for me. If we're sure that the local.php file will be rewrite by the admin interface of dokuwiki, in that case I think that we shouldn't overwrite this file at each upgrade. That would means overwrite any user config.

You're right. https://www.dokuwiki.org/install:upgrade
I will add the same test as for the file `acl.auth.php` to avoid overwrite and copy it if does not exist beforehand.

Reply:
Sounds good to me

Reply:
> I think that solution2 is more flexible as the user should not modify authtype usually and that only "power users" will want this feature.

Considering only power users would change that setting, wouldn't it be better to leave it in `local.protected.php` ? So that a "casual" user will not modify it, even by mistake, but a power user will be able to modify it by ssh.
It could even become a part of the `config-panel` feature.

Reply:
> I was not aware of the possibility of the config-panel before the release of Yunohost 3.1.

It's a new and experimental feature, but it would be the better solution here, even if it's not fully available for all users right now.
Using a setting like `is_internal_users` is a good idea. It seems to me like the better thing to do.

**Answers time**

1. You don't have to, and it would even be better to not. You can add any setting you want to an app without adding it to the manifest. The manifest is used mainly for the install panel shown to the user.
2. Set `is_internal_users` to false and explain why. To prevent any further questions about it.
3. You can find a real use case with this PR for etherpad https://github.com/YunoHost-Apps/etherpad_mypads_ynh/pull/49. Maybe it would be more clear than the original PR.
We don't have any documentation yet, because those features are still experimental.
4. I would say a `"help"` entry to explain it in the config-panel and an error with a warning.
Also, an `action` to switch from private to public and the other way. So an user can do whatever he wants.
5. If you can't install a plugin automatically by CLI (Don't know myself, never tried this app), same as 4.
6. Nice ! A documentation is always welcome !

Reply:
I am reading the PR https://github.com/YunoHost-Apps/etherpad_mypads_ynh/pull/49 and trying to "translate" it to the use case discussed here.
1. When should I use an `actions` script instead of doing everything in `config` ?
    - You don't use an "action script" for "is_public" [here](https://github.com/YunoHost-Apps/etherpad_mypads_ynh/pull/49/files#diff-fbc1eef3364b71e1e9debe976c54d7eaR232-245) ==> It is redundant with `install` and `upgrade` and cannot be called from CLI if I understand it correctly
    - Here, you use a dedicated script : https://github.com/YunoHost-Apps/wordpress_ynh/pull/47

I am a little bit confused by all theses possibilities and trying to sort them out takes time. I will leave time (and other volunteers) to smooth out theses brand new functionalities and provide documentation later.

It makes me wondering : Is this kind of feedback useful for the project?

Reply:
It doesn't really matter if you use an action or a config.
There's nothing specific yet, I would say anyway that actions are used for a CLI action. And config are more for all is about a config file for the app.

In case of public/private, even if we have (I have ?) to be consistent and propose a standard way to do it, there's nothing decided yet. Maybe an action would be better to do that.